### PR TITLE
[type:feat]The sync-data-http method adapts to the namespace transformation

### PR DESCRIPTION
**Pull request by xcsnx**
Make sure that:

- [ ] You have read the [contribution guidelines](https://shenyu.apache.org/community/contributor-guide).
- [ ] You submit test cases (unit or integration tests) that back your changes.
- [ ] Your local test passed `./mvnw clean install -Dmaven.javadoc.skip=true`.

### CODE REVIEW

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/http/HttpLongPollingDataChangedListener.java [hunk 1]
```diff
@@ -139,46 +144,47 @@ public void doLongPolling(final HttpServletRequest request, final HttpServletRes
...
         asyncContext.setTimeout(0L);
         // block client's thread.
-        scheduler.execute(new LongPollingClient(asyncContext, clientIp, HttpConstants.SERVER_MAX_HOLD_TIMEOUT));
+        scheduler.execute(new LongPollingClient(asyncContext, clientIp, HttpConstants.SERVER_MAX_HOLD_TIMEOUT, namespaceId));
     }
 
     @Override
     protected void afterAppAuthChanged(final List<AppAuthData> changed, final DataEventTypeEnum eventType) {
-        scheduler.execute(new DataChangeTask(ConfigGroupEnum.APP_AUTH));
+        scheduler.execute(new DataChangeTask(ConfigGroupEnum.APP_AUTH, ""));
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/http/HttpLongPollingDataChangedListener.java:152]: why empty string?

**Review by moremind — CHANGES_REQUESTED** [review]

**Comment by moremind** [comment]
I think http long pull just add shenyu-bootstrap<shenyu.namespaceId> as prefix for plugin,selector,rule,metadata and so on.

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/AbstractDataChangedListener.java [hunk 2]
```diff
@@ -97,14 +100,22 @@ public abstract class AbstractDataChangedListener implements DataChangedListener
...
      * fetch configuration from cache.
      *
-     * @param groupKey the group key
+     * @param groupKey    the group key
+     * @param namespaceId the namespaceId
      * @return the configuration data
      */
-    public ConfigData<?> fetchConfig(final ConfigGroupEnum groupKey) {
+    public ConfigData<?> fetchConfig(final ConfigGroupEnum groupKey, final String namespaceId) {
+        //todo:[Namespace] Currently, only plugin data is compatible with namespace, while other data is waiting for modification
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/AbstractDataChangedListener.java:114]: why not remove this?

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/AbstractDataChangedListener.java [hunk 3]
```diff
@@ -166,19 +177,20 @@ public void onPluginChanged(final List<PluginData> changed, final DataEventTypeE
         if (CollectionUtils.isEmpty(changed)) {
             return;
         }
-        this.updatePluginCache();
-        this.afterPluginChanged(changed, eventType);
+        Optional<String> namespaceId = changed.stream().map(PluginData::getNamespaceId).findFirst();
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/AbstractDataChangedListener.java:180]: you should judge Optional.isEmpty??

Comment (moremind) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/AbstractDataChangedListener.java:180]: avoid npe

**Review by moremind — CHANGES_REQUESTED** [review]

File: shenyu-admin/src/test/java/org/apache/shenyu/admin/controller/ConfigControllerTest.java [hunk 4]
```diff
@@ -71,12 +71,12 @@ public void setUp() {
     public void testFetchConfigs() throws Exception {
         // Configure HttpLongPollingDataChangedListener.fetchConfig(...).
         final ConfigData<?> configData = new ConfigData<>("md5-value1", 0L, Collections.emptyList());
-        doReturn(configData).when(mockLongPollingListener).fetchConfig(ConfigGroupEnum.APP_AUTH);
+        doReturn(configData).when(mockLongPollingListener).fetchConfig(ConfigGroupEnum.APP_AUTH, "649330b6-c2d7-4edc-be8e-8a54df9eb385");
```
Comment (moremind) [inline shenyu-admin/src/test/java/org/apache/shenyu/admin/controller/ConfigControllerTest.java:74]: not use magic value

File: shenyu-admin/src/test/java/org/apache/shenyu/admin/listener/AbstractDataChangedListenerTest.java [hunk 5]
```diff
@@ -70,6 +73,10 @@ public final class AbstractDataChangedListenerTest {
 
     private DiscoveryUpstreamService discoveryUpstreamService;
 
+    private final String namespaceId = "649330b6-c2d7-4edc-be8e-8a54df9eb385";
```
Comment (moremind) [inline shenyu-admin/src/test/java/org/apache/shenyu/admin/listener/AbstractDataChangedListenerTest.java:76]: not use magic value,, use common value

**Review by moremind — CHANGES_REQUESTED** [review]

File: shenyu-common/src/main/java/org/apache/shenyu/common/config/ShenyuConfig.java [hunk 6]
```diff
@@ -70,7 +70,9 @@ public class ShenyuConfig {
     private SpringCloudCacheConfig springCloudCache = new SpringCloudCacheConfig();
     
     private AlertConfig alert = new AlertConfig();
-    
+
+    private String namespaceId = "";
```
Comment (moremind) [inline shenyu-common/src/main/java/org/apache/shenyu/common/config/ShenyuConfig.java:74]: can set empty string

**Review by moremind — CHANGES_REQUESTED** [review]

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/http/HttpLongPollingDataChangedListener.java [hunk 7]
```diff
@@ -94,7 +98,7 @@ public class HttpLongPollingDataChangedListener extends AbstractDataChangedListe
      * @param httpSyncProperties the HttpSyncProperties
      */
     public HttpLongPollingDataChangedListener(final HttpSyncProperties httpSyncProperties) {
-        this.clients = new ArrayBlockingQueue<>(1024);
+        this.clientsMap = new CompactHashMap<>();
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/http/HttpLongPollingDataChangedListener.java:101]: why use this? why not hashmap?

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/http/HttpLongPollingDataChangedListener.java [hunk 8]
```diff
@@ -187,7 +193,8 @@ private List<ConfigGroupEnum> compareChangedGroup(final HttpServletRequest reque
             }
             String clientMd5 = params[0];
             long clientModifyTime = NumberUtils.toLong(params[1]);
-            ConfigDataCache serverCache = CACHE.get(group.name());
+
+            ConfigDataCache serverCache = CACHE.get(namespaceId + "_" + group.name());
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/http/HttpLongPollingDataChangedListener.java:197]: you can new a method to change this, may `buildCacheKey`

**Review by moremind — CHANGES_REQUESTED** [review]

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/http/HttpLongPollingDataChangedListener.java (hunk 7 above)
Comment (xcsnx) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/http/HttpLongPollingDataChangedListener.java:101]: Indeed, I will improve

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/NamespaceServiceImpl.java [hunk 9]
```diff
@@ -99,10 +99,17 @@ public NamespaceVO findById(final String namespaceId) {
 
     @Override
     public List<NamespaceVO> list(final String name) {
-        List<NamespaceDO> namespaceDOS = namespaceMapper.selectAll(name);
+        List<NamespaceDO> namespaceDOS = namespaceMapper.selectAllByName(name);
         return namespaceDOS.stream().map(NamespaceTransfer.INSTANCE::mapToVo).collect(Collectors.toList());
     }
 
+    @Override
+    public List<NamespaceVO> list() {
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/NamespaceServiceImpl.java:107]: rename listAll

File: shenyu-sync-data-center/shenyu-sync-data-http/src/main/java/org/apache/shenyu/sync/data/http/HttpSyncDataService.java [hunk 10]
```diff
@@ -188,6 +194,7 @@ private void doLongPolling(final String server) {
                 params.put(group.name(), Lists.newArrayList(value));
             }
         }
+        params.put("namespaceId", Lists.newArrayList(shenyuConfig.getNamespace()));
```
Comment (moremind) [inline shenyu-sync-data-center/shenyu-sync-data-http/src/main/java/org/apache/shenyu/sync/data/http/HttpSyncDataService.java:197]: not a list

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ConfigController.java [hunk 11]
```diff
@@ -35,36 +40,50 @@
...
      */
     @GetMapping("/fetch")
-    public ShenyuAdminResult fetchConfigs(@NotNull final String[] groupKeys) {
+    public ShenyuAdminResult fetchConfigs(@NotNull final String[] groupKeys, final String namespaceIdParams) {
+        String namespaceId = namespaceIdParams;
+        NamespaceVO existNamespace = namespaceService.findById(namespaceId);
+        if (StringUtils.isNotEmpty(namespaceId) && ObjectUtils.isEmpty(existNamespace)) {
+            throw new ShenyuAdminException("namespaceId is not exist");
+        }
+        if (StringUtils.isEmpty(namespaceId)) {
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ConfigController.java:76]: remove this, the param is null, you just throw exception

Comment (moremind) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ConfigController.java:76]: you should refactor this code

**Review by moremind — CHANGES_REQUESTED** [review]

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/NamespaceServiceImpl.java (hunk 9 above)
Comment (xcsnx) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/NamespaceServiceImpl.java:107]: ok

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/config/HttpLongPollingSyncConfiguration.java [hunk 12]
```diff
@@ -43,4 +45,16 @@ public HttpLongPollingDataChangedListener httpLongPollingDataChangedListener(fin
...
+     * configController.
+     *
+     * @param httpLongPollingDataChangedListener httpLongPollingDataChangedListener
+     * @param namespaceService namespaceService
+     * @return {@link ConfigController}
+     */
+    @Bean
+    @ConditionalOnMissingBean(ConfigController.class)
+    public ConfigController configController(final HttpLongPollingDataChangedListener httpLongPollingDataChangedListener, final NamespaceService namespaceService) {
+        return new ConfigController(httpLongPollingDataChangedListener, namespaceService);
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/config/HttpLongPollingSyncConfiguration.java:58]: delete this

**Review by moremind — CHANGES_REQUESTED** [review]

**Review by moremind — CHANGES_REQUESTED** [review]

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/config/HttpLongPollingSyncConfiguration.java (hunk 12 above)
Comment (xcsnx) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/config/HttpLongPollingSyncConfiguration.java:58]: ok

**Review by moremind — APPROVED** [review]

**Pull request merged**